Add endpoints info data and information about CQRS (#75)
many-protocol/specification: spec/protocol/endpoints.adoc

= MANY Endpoints

> A server exposes endpoints that a message must target.

Reviewer: Endpoints are methods exposed by servers that a client can trigger to perform an operation.
A client can trigger an endpoint by sending the server a correctly-formatted `bytestring` payload. (Add a link to the spec/CDDL)
The server will reply a `bytestring` to the client after the operation completes. The reply may contain the result of the operation or an error if the operation failed to execute properly.
Endpoints follow the `module.endpoint` naming convention, e.g., `ledger.send`.

Or something like that.

Author: >  Endpoints follow the module.endpoint naming convention, e.g., ledger.send.

It should, but it doesn't have to, especially for endpoints that don't care about the spec attributes. If you're making endpoints for a private server or something, you don't need to do that.

Author: I'm in the process of rewording, I think, but we can also improve on wording later on.

> Endpoints receive a single bytestring payload, and optionally return either an error or a bytestring payload.

Reviewer: They always return something, but that something can be empty.

== Command Query Responsibility Segregation

MANY separates messages into two kinds;

> 1. Query (or read) which access data in read-only mode.

Reviewer: ```suggestion
1. Query: Access data in read-only mode.
```

>    The app state hash should not be changed after running a query, and any queued command calls should be resolved or paused (commands SHOULD NOT be in the middle of executing).

Reviewer: The 2nd part is not clear.

I think you mean that some data MUST NOT be accessed in both read AND write modes simultaneously, i.e., NO race condition.

Author: Queries can be executed with race conditions because they cannot change the memory. Commands cannot. I'll try to clarify.

>    The app state hash should not be changed after running a query, and any queued command calls should be resolved or paused (commands SHOULD NOT be in the middle of executing).

Reviewer: ```suggestion
   Executing a query should NOT change the application state hash. Any queued command calls should be resolved or paused (commands SHOULD NOT be in the middle of executing).
```

   Query results can be cached on the state hash, and they can also be executed at the same time since the state does not change.

> 2. Command (or update) which perform changes to the data.

Reviewer: ```suggestion
2. Command: Access data in write mode.
```

>    This does not necessarily mean that the app state must change, just that it can.

Reviewer: Mention that the application state hash MIGHT change after executing a command.

>    This does not necessarily mean that the app state must change, just that it can.

Reviewer: Remove

> An endpoint needs to be either.

Reviewer: Remove

> A server SHOULD use commands for calls that it isn't sure can update the state (for example, conditionally).

Reviewer: Give a concrete example

= MANY Endpoints

Endpoints are methods exposed by servers that a client can call with a single bytestring payload.
The server will reply to the client in a timely matter footnote:[If this cannot be done synchronously, the server MUST still respond with a response, see xref:../../attributes/network/8_async.adoc[Async Attribute].].
The reply MAY contain the result of the operation or an error if the operation failed to execute properly.

== Command Query Responsibility Segregation

MANY separates messages into two kinds;

1. Query: access data in read-only mode.
   Executing a query SHOULD NOT change the application state hash.

>    Any queued command calls should be resolved or paused (commands SHOULD NOT be in the middle of executing) prior to executing a query.

Reviewer: ```suggestion
   Any queued command calls should be resolved or paused before executing a query, i.e., the server SHOULD NOT execute a query while a command is running.
```

nit: I still don't like the `should be resolved or paused`

Author: Tried to reword removing that. We can wordsmith more later.
